### PR TITLE
[ABA Impact] - Latest formula and tests

### LINKED ISSUE #1495: 🎶 ABA Test: Sell and buy back to calculate slippage
**Description**
Xdai, has many USD estimations missing. This is because, ther USD oracles are for mainnet.

This ABA test (A->B->A test) cosist on simulate doing a trade back and forth the asset, and see how much you get back from your original.

In a naive implementation we could say that slippage is half of that.
i.e.
10 DAI gives me 9 USDC
9 USDC gives me 8 DAI
100 * (10-8)/10/2  ----> 10%

However, dividing by two wouldn't give us the estimated slippage.

The idea is to solve it in the following way:
> Note: It has a small error factoring `(p-ps)(1-s))` (one of the signs are wrong)
image

Context:
https://gnosisinc.slack.com/archives/CPZA1AGKY/p1632220484032000?thread_ts=1632210034.021000&cid=CPZA1AGKY

### PR DESCRIPTION
**Pull request by W3stside**
Closes #1495

Adds the latest formula by Anxo written here: https://github.com/gnosis/cowswap/issues/1495

The tests are in utils/computePriceImpact

Test: as usual

This is merging into base test logic so has no loaders etc for the moment

### CODE REVIEW

File: src/custom/utils/price.ts [hunk 1]
```diff
@@ -296,28 +297,101 @@ export function getValidParams(params: PriceQuoteParams) {
...
+    const Pab = middleValueBn.div(initialValueBn)
+    const Pba = finalValueBn.div(middleValueBn)
+
+    Psqrt = Pab.times(Pba).sqrt()
+  } else {
+    // Pab = abOut/abIn
+    // Pba = baOut/baIn[abIn]
+    const initialValueBn = new BigNumberJs(initialValue)
+    const middleValueBn = new BigNumberJs(middleValue)
+    const finalValueBn = new BigNumberJs(finalValue)
```
Comment (alfetopito) [inline src/custom/utils/price.ts:380]: Nitpick: the instantiation of iv, mv and fv can be done outside of the if

File: src/custom/utils/price.ts [hunk 2]
```diff
@@ -296,28 +297,101 @@ export function getValidParams(params: PriceQuoteParams) {
...
+    const Pab = initialValueBn.div(middleValueBn)
+    // 2400/122
+    const Pba = finalValueBn.div(middleValueBn)
+
+    Psqrt = Pab.div(Pba).sqrt()
+  }
 
-  console.debug(`[calculateFallbackPriceImpact]::${priceImpactPercentage.toSignificant(2)}%`)
+  const [numerator1, denominator1] = ONE_BIG_NUMBER.minus(Psqrt).toFraction()
+  const [numerator2, denominator2] = ONE_BIG_NUMBER.plus(Psqrt).toFraction()
```
Comment (alfetopito) [inline src/custom/utils/price.ts:391]: Dumb question: why convert to fraction?

Also, could this ever be negative?

File: src/custom/utils/price.ts [hunk 3]
```diff
@@ -296,28 +297,101 @@ export function getValidParams(params: PriceQuoteParams) {
...
 
-  return priceImpactPercentage
+  // is response within [0,1] range?
+  return _isWithinRange(numerator1, denominator1)
```
Comment (alfetopito) [inline src/custom/utils/price.ts:394]: Is it possible that both are within range?
How would we pick one in that case?

File: src/custom/utils/price.ts [hunk 4]
```diff
@@ -296,28 +297,101 @@ export function getValidParams(params: PriceQuoteParams) {
...
+// -------------------------------------------
+// EXAMPLE [BUY]
+// -------------------------------------------
+// 2970B > 122A // 122A >> 2400B
+// --> 2970/122 = 24.34
+// --> 2400/122 = 19.67
+// -------------------------------------------
+// SIDE 1 - BUY
+// -------------------------------------------
+// S = 1 - sqrt( 24.34/19.67 ) = 0.2516685226 --> -0.109 --> -10.9%, valid cause is a number between 0 and 1
```
Comment (alfetopito) [inline src/custom/utils/price.ts:345]: `-0.109` is not between `0` and `1` :thinking: 
Unless you mean the modulus?

**Review by alfetopito — COMMENTED** [review]
Finished reviewing the code. A bit confused by the concepts but looks ok.
Will test the behaviour next

**Comment by alfetopito** [comment]
Trying it out on xDai:

# wxdai - 1inch
## Sell
Screenshot from 2021-11-30 11-43-00

## Buy
Screenshot from 2021-11-30 11-42-37

# wxdai - usdc

Sell is fine

# Buy
Estimate is 0% which is not really accurrate
Screenshot from 2021-11-30 11-45-36

**Comment by alfetopito** [comment]
Another pair weth-dai and vice-versa.

Everything seems off, ~even when there's an usd amount~ (this is because it's always on on purpose)

Screenshot from 2021-11-30 11-50-30
Screenshot from 2021-11-30 11-49-52
Screenshot from 2021-11-30 11-49-23
Screenshot from 2021-11-30 11-48-54

**Comment by alfetopito** [comment]
aave-dai and vice-versa, compared with honeyswap
Screenshot from 2021-11-30 11-58-40
Screenshot from 2021-11-30 11-58-05
Screenshot from 2021-11-30 11-57-43
Screenshot from 2021-11-30 11-56-44

**Review by anxolin — COMMENTED** [review]
There’s something weird…

In `mainnet`, for the AB trade:

* I sold 1 ETH for DAI
* it estimated I received  4665.0155 DAI
* It gave me -2.67%

However, if I then input the BA trade
* Sell 4665.0155 DAI for ETH
* it estimated I received   1.0097 WETH
* It showed me -2.6%

Those % are too big, look in this example below, this is mainnet, it should be a liquid market, i shouldnt be loosing 2.5

image

In term of USD estimation is not so severe impact. I guess there must be an issue with the calculation, or with this method.

One thing is comming to mind that can be a partial culprit of this is that if u trade back and forth then u pay 2 liquidity provider fees (that's the price u get), so u might start with `-0.3%` already, however this doesn't explain this big difference

Maybe, we should focus on reasoning our cases with the unit tests?

File: src/custom/utils/calculatePriceImpact.test.ts [hunk 5]
```diff
@@ -1,53 +1,67 @@
...
-  const abOut = CurrencyAmount.fromRawAmount(DAI_MAINNET, AB_OUT)
-
   describe('[SELL] WETH --> DAI', () => {
     it('A > B > A SELL return proper price impact', () => {
-      // GIVEN a 1 WETH >> 1000 DAI AB Trade
-      // GIVEN a 1000 DAI >> 0.5 WETH BA Trade
-      const BA_OUT = parseUnits('0.5', WETH_MAINNET.decimals).toString()
-      const baOut = CurrencyAmount.fromRawAmount(WETH_MAINNET, BA_OUT)
+      // GIVEN a 100 WETH >> 80 DAI AB Trade
+      const abInParsed = parseUnits('100', WETH_MAINNET.decimals).toString()
```
Comment (anxolin) [inline src/custom/utils/calculatePriceImpact.test.ts:13]: All this test follow the same structure


probably nice to define all cases in a simpler way (a const with the minimal data:
```
const ABA_CASES = [
   {
       soldAmount: 100, 
       receivedAmount: 80,
       recoveredAmount: 56,
       result: '25'
   }, ...
]
```

This way, is easier to review, create new cases, and u can even create `it` cases with dynamic names, so you know exactly the case that succeeds and/or fails

File: src/custom/utils/price.ts (hunk 4 above)
Comment (anxolin) [reply src/custom/utils/price.ts:345]: The 0...1 thingy is something I saw experimentally, but is likely not the way to choose the right solution

not sure, we should do more examples.

Comment (anxolin) [reply src/custom/utils/price.ts:345]: Actually, thinking a bit more on this, I tried to justify the logic to pick the sign of the formula.

After a lot of thinking, I arrived to a conclusion, although this doesn't mean my conclusion is correct  😂

TLDR: We should always take the `MINUS` sign.

If you want the full explanation why I claim this:

IMG_3551

IMG_3552

**Comment by elena-zh** [comment]
Hey, a stupid question: when we do a BA trade, which exactly amount we take into account: Receive or Receive incl. fee?
image

Then again, seems a bit strange to me to see almost 0 price impact for a buy order. See the comparison with Prod: 
buy order

Just not mention: when impact is loading, it may show for a while too odd value (might frighten someone)
faced this

In a while it is changed to an appropriate one. I believe, it will be fixed by a loading indicator in one of related PRs

**Comment by W3stside** [comment]
closing for now
